contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,450 @@
+to somehow ask for all Secrets attached to all pods bound to a given node in a
+single LIST or WATCH request). However, the design has to be consistent with
+authorization of other types of requests (in particular GETs).
+
+ To solve this problem, we propose to introduce the idea of `bulk watch ` and
+`bulk list ` (and in the future other bulk operations). This idea was already
+appearing in the community and now we have good usecase to proceed with it.
+
+ Once a bulk watch is set up, we also need to periodically verify its ACLs.
+Whenever a user (in our case Kubelet) loses access to a given resource, the
+watch should be closed within some bounded time. the rationaly behind this
+requirement is that by using (bulk) watch we can't regress the current behavior
+where a polling Kubelet would eventually be unable to access a secret.
+
+ That said, periodic verification of ACLs isn't specific to bulk watch and
+needs to be solved also in `reqular` watch (e.g. user watching just a single
+secret may also lose access to it and such watch should also be closed in this
+case). So this requirement is common for both regular and bulk watch. We
+just need to solve this problem on low enough level that would allow us to
+reuse the same mechanism in both cases - we will solve it by sending an error
+event to the watcher and then just closing this particular watch.
+
+ At the high level, we would like the API to be generic enough so that it will
+be useful for many different usecases, not just this particular one (another
+example may be a controller that needs to deal with just subset of namespaces).
+As a result, below we are describing requirements that the end solution has to
+meet to satisfy our needs
+- a single bulk requests has to support multiple resource types (e.g. get a
+node and all pods associated with it)
+- if the bulk requests filters asks only for object of a single resource type,
+the returned result ideally should be compatible with what is returned from
+the regular API
+- the API has to support aggregation if different resource types are served
+by different apiservers
+- clients has to be able to alter their watch subscribtions incrementally (it
+may not be implemented in the initial version though, but has to be designed)
+
+
+# Detailed design
+
+ As stated in above requirements, we need to make bulk operations work across
+different resource types (e.g. watch pod P and secret S within a single watch
+call). To clarify, the initial implementation doesn't need to support it, but
+it has to be possible to extend the implementation without changing the API.
+Spanning multiple resources, resources types or conditions will be more and more
+important for large number of watches. As an example, federation will be adding
+watches for every type it federates. With that in mind, bypassing aggregation
+at the resource type level and going to aggregation over objects with different
```

Comment:
Yeah, to be able to have a server (not a specific endpoint) efficiently handle multi watch.  It's the obvious first point of perf improvement - push the selector down to the storage level.  We've talked about that for other types of operations (bulk namespace fetch)